### PR TITLE
[HOP][print] Add HOP subclass for printing

### PR DESCRIPTION
**Pull request by fxdawnn**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167016
* #166920
* __->__ #166660

### CODE REVIEW

File: test/higher_order_ops/test_print.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class TestHopPrint(TestCase):
+    def test_base_print(self):
+        # class M(torch.nn.Module):
+        #     def forward(self, x):
+        #         torch._higher_order_ops.print("moo")
+        #         res = x + x
+        #         torch._higher_order_ops.print(("moo")
+        #         return (res,)
```
Comment (angelayi) [inline test/higher_order_ops/test_print.py]: delete comments?

File: torch/_higher_order_ops/hop_print.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    def __init__(self):
+        super().__init__("hop_print")
+        self._print_str = None
+
+    def __call__(self, *args, **kwargs):
+        if self._print_str is None:
+            self._print_str = self._get_print_str()
+        print(self._print_str)
+        return self._fn(*args, **kwargs)
+
```
Comment (angelayi) [inline torch/_higher_order_ops/hop_print.py]: can you follow what is done in [other HOO](https://github.com/pytorch/pytorch/blob/52db60170dd192afa48233f03574a8d8bcd39d35/torch/_higher_order_ops/executorch_call_delegate.py#L28-L33) where we have an implementation for the [CompositeExplicitAutograd](https://github.com/pytorch/pytorch/blob/52db60170dd192afa48233f03574a8d8bcd39d35/torch/_higher_order_ops/executorch_call_delegate.py#L76-L90) dispatch key?

File: torch/_higher_order_ops/hop_print.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import torch.utils._pytree as pytree
+from torch._ops import HigherOrderOperator
+
+
+class Print(HigherOrderOperator):
+    def __init__(self):
+        super().__init__("hop_print")
+        self._print_str = None
+
+    def __call__(self, *args, **kwargs):
```
Comment (angelayi) [inline torch/_higher_order_ops/hop_print.py]: we want the signature to be like format fmt string + args + kwargs so that users can do `torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)`

File: test/higher_order_ops/test_print.py (hunk 1 above)
Comment (fxdawnn) [reply test/higher_order_ops/test_print.py]: deleted.

File: torch/_higher_order_ops/hop_print.py (hunk 2 above)
Comment (fxdawnn) [reply torch/_higher_order_ops/hop_print.py]: Sounds great! Fixed.

File: torch/_higher_order_ops/hop_print.py (hunk 3 above)
Comment (fxdawnn) [reply torch/_higher_order_ops/hop_print.py]: Amazing! Fixed and added test set.

File: torch/_higher_order_ops/print.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import builtins
+
+import torch
+import torch.utils._pytree as pytree
+from torch._ops import HigherOrderOperator
+
+
+# This is a functional version of torch.print to enable format printing
+# through calling such as
+# torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
```
Comment (angelayi) [inline torch/_higher_order_ops/print.py]: can you write the docstring like [this](https://fburl.com/code/xdxglcup)?

File: torch/_higher_order_ops/print.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            tuple(map_types.keys()),
+            lambda a: map_types[type(a)](a),
+            kwargs,
+            lambda a: isinstance(a, tuple(map_types.keys())),
+        )
+        print_str = format_str.format(**new_kwargs)
+    else:
+        print_str = format_str
+    # Use built-in print to avoid recursion with the HOP print
+    builtins.print(print_str)
```
Comment (angelayi) [inline torch/_higher_order_ops/print.py]: nit
```suggestion
      new_kwargs = pytree.tree_map_only(
          tuple(map_types.keys()),
          lambda a: map_types[type(a)](a),
          kwargs,
          lambda a: isinstance(a, tuple(map_types.keys())),
      )

    # Use built-in print to avoid recursion with the HOP print
    builtins.print(format_str.format(new_kwargs))
```

File: torch/testing/_internal/hop_db.py [hunk 6]
```diff
@@ -80,17 +80,17 @@ def f2(x, y0, y1):
...
+    "executorch_call_delegate",
+    "wrap",
+    "wrap_with_set_grad_enabled",
+    "auto_functionalized_v2",
+    "associative_scan",
+    "flat_apply",  # is WIP, doesn't pass any of the tests yet
+    "wrap_with_autocast",
+    "wrap_activation_checkpoint",
+    "run_const_graph",
+    "auto_functionalized",
```
Comment (angelayi) [inline torch/testing/_internal/hop_db.py]: nit: can you remove the formatting changes in this file?

Comment (fxdawnn) [reply torch/testing/_internal/hop_db.py]: somewhat the linter won't allow me at all.

File: torch/_higher_order_ops/print.py (hunk 4 above)
Comment (fxdawnn) [reply torch/_higher_order_ops/print.py]: just updated. Thanks!

File: torch/_higher_order_ops/print.py (hunk 5 above)
Comment (angelayi) [reply torch/_higher_order_ops/print.py]: bump :P

Comment (fxdawnn) [reply torch/_higher_order_ops/print.py]: finally.. fixed lol!

File: torch/testing/_internal/hop_db.py (hunk 6 above)
Comment (fxdawnn) [reply torch/testing/_internal/hop_db.py]: fix formatOnSave and fixed

File: torch/_higher_order_ops/print.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    This Higher Order Operator (HOP) provides a functional version of print for use in PyTorch graphs.
+    It enables format printing with named arguments, e.g., torch._higher_order_ops.print("moo {x} {y}", x=1, y=2).
+
+    This HOP enables printing without causing graph break.
+    """
+
+    def __init__(self) -> None:
+        super().__init__("print")
+
+    def __call__(self, format_str: str, **kwargs: object) -> object:
```
Comment (angelayi) [inline torch/_higher_order_ops/print.py:21]: can you assert that format_str is a str here?

Comment (fxdawnn) [reply torch/_higher_order_ops/print.py:21]: yes! Fixed

**Review by angelayi — APPROVED** [review]

**Review by anijain2305 — APPROVED** [review]
